### ISSUE TITLE
Make _quicksort work with raw memory addresses

### ISSUE THREAD

**Issue by CodeSandwich** [OpenZeppelin/openzeppelin-contracts#4882]
**🧐 Motivation**
Go fast.

**📝 Details**
Function `sort` from `utils/Arrays.sol` uses a private function `_quickSort`, which works on the array pointer and two indexes. It makes multiple array accesses, and for each of them it needs to calculate the offset of the actual data (`+32`) and the position of the array item (`+32*idx`).

An alternative would be to make `_quickSort` accept 2 raw memory pointers, and work with them. `sort` would call `assembly { let ptr := array } _quicksort(ptr + 32, ptr + 32 + array.length * 32>`. Instead of performing `idx++` to switch to the next item, `_quickSort` would do identically priced `ptr+=32`. Accessing the memory would be much cheaper, because it would be plain `mload`, without the index to pointer conversion. This approach would probably have less logic in assembly than the current one, because it would be limited to simple `mload`s and `mstore`s, probably just 2 one-line helper functions. The code would be comparably unsafe and would require similar expertise to understand when compared the current implementation, which does unchecked memory array operations anyway.

WDYT?

**Comment by Amxx** [comment]
That is an interresting idea. 

However, I tried doing
```solidity
    function sort(uint256[] memory array) internal pure returns (uint256[] memory) {
        uint256 start;
        uint256 end;
        assembly {
            start := add(array, 0x20)
            end := add(start, mul(mload(array), 0x20))
        }
        _quickSort(start, end);
        return array;
    }

    function _quickSort(uint256 start, uint256 end) private pure {
        unchecked {
            // Can't overflow given `i <= j`
            if (end - start < 0x40) return;

            // Use first element as pivot
            bytes32 pivot = _mload(start);
            // Position where the pivot should be at the end of the loop
            uint256 pos = start;

            for (uint256 it = start + 0x20; it < end; it += 0x20) {
                if (uint256(_mload(it)) < uint256(pivot)) {
                    // If array[k] is smaller than the pivot, we increment the position of the pivot and move array[k]
                    // there.
                    pos += 0x20;
                    _swap(pos, it);
                }
            }

            _swap(start, pos); // Swap pivot into place
            _quickSort(start, pos); // Sort the left side of the pivot
            _quickSort(pos + 0x20, end); // Sort the right side of the pivot
        }
    }

    function _mload(uint256 ptr) private pure returns (bytes32 value) {
        assembly { value := mload(ptr) }
    }

    function _mstore(uint256 ptr, bytes32 value) private pure {
        assembly { mstore(ptr, value) }
    }

    function _swap(uint256 ptr1, uint256 ptr2) private pure {
        bytes32 val1 = _mload(ptr1);
        bytes32 val2 = _mload(ptr2);
        _mstore(ptr1, val2);
        _mstore(ptr2, val1);
    }
```
and the gaz cost increassed. 

We would be open to any refactor of the code that improves performance, including ones that modify the private functions.

**Comment by CodeSandwich** [comment]
Thanks for looking into it and benchmarking this approach so quickly! I'm surprised by the results, but if you would be interested in merging a refactoring like this, I'll look into it, I must be missing something.

**Comment by Amxx** [comment]
My benchmarking was very fast. Maybe there is more to it.

Also, there may be a hidden cost in 
```solidity
    function _swap(uint256 ptr1, uint256 ptr2) private pure {
        bytes32 val1 = _mload(ptr1);
        bytes32 val2 = _mload(ptr2);
        _mstore(ptr1, val2);
        _mstore(ptr2, val1);
    }
```

Maybe doing that full assembly would improve things

**Comment by CodeSandwich** [comment]
I did a very quick benchmark of your code with solc 0.8.24, via-IR and 1B optimizer runs. On my side the results are **very** good:

| | gas old | gas new |
| - | - | - |
|100 random | 143,861 | 94,868 |
| 100 same | 626,457 | 491,535|
|100 sorted | 626,457 | 491,535 |
|100 reverse sorted | 923,957 | 611,535 |

**Comment by Amxx** [comment]
The configuration of the optimizer is something I'm concerned with.
Depending on how the optimizer is configured, for changes may be good, but they would be bad if the configuration was different.

You're context (via-ir) + 1B runs, is very specific, and IMO does not match the "common" config of users. I am under the impression that the most common config is "optimizer enabled + 200 runs + viaIR disabled". I believe we should optimize or code for that. This choice may lead us to make decisions that are "bad" for users that use different config ... but we honestly cannot release multiple version of the library depending on the targetted compiler.

All that being said, I will try to replicate you number, and dig deeper, early next week !

**Comment by CodeSandwich** [comment]
That's a good observation, with the legacy pipeline and 200 runs the results are mixed, and often the current approach is cheaper.

**Comment by Amxx** [comment]
I rewrote the `_swap` with pointers to be pure assembly, and is saves A LOT of gas. I'm guessing `_mload` and `_mstore` are not inlined and that is the issue. Creating a PR

**Comment by Amxx** [comment]
Implemented in #4883
